src/encoded/reports/metadata.py

```diff
@@ -0,0 +1,365 @@
+
+    def _get_audit_data(self, grouped_audits_for_file, grouped_other_audits):
+        return {
+            audit_column: ', '.join(
+                set(
+                    grouped_audits_for_file.get(audit_type, [])
+                    + grouped_other_audits.get(audit_type, [])
+                )
+            ) for audit_type, audit_column in METADATA_AUDIT_TO_AUDIT_COLUMN_MAPPING
+        }
+
+    def _output_sorted_row(self, experiment_data, file_data):
+        row = []
+        for column in self.header:
+            row.append(
+                file_data.get(
+                    column,
+                    experiment_data.get(column)
+                )
+            )
+        return row
+
+    def _generate_rows(self):
+        yield self.csv.writerow(self.header)
+        for experiment in self._get_search_results_generator()['@graph']:
+            if not experiment.get('files', []):
+                continue
+            grouped_file_audits, grouped_other_audits = group_audits_by_files_and_type(
+                experiment.get('audit', {})
+            )
+            experiment_data = self._get_experiment_data(experiment)
+            for file_ in experiment.get('files', []):
+                if self._should_not_report_file(file_):
+                    continue
+                file_data = self._get_file_data(file_)
+                audit_data = self._get_audit_data(
+                    grouped_file_audits.get(file_.get('@id'), {}),
+                    grouped_other_audits
+                )
+                file_data.update(audit_data)
+                yield self.csv.writerow(
+                    self._output_sorted_row(experiment_data, file_data)
+                )
+
+    def generate(self):
+        self._validate_request()
+        self._initialize_report()
+        self._build_params()
+        return Response(
+             content_type='text/tsv',
+             app_iter=self._generate_rows(),
+             content_disposition='attachment;filename=metadata.tsv'
+        )
+
+
+class CSVGenerator:
+
+    def __init__(self, delimiter='\t', lineterminator='\n'):
+        self.writer = csv.writer(
+            self,
+            delimiter=delimiter,
+            lineterminator=lineterminator
+        )
+
+    def writerow(self, row):
+        self.writer.writerow(row)
+        return self.row
+
+    def write(self, row):
+        self.row = row.encode('utf-8')
+
+
+def _get_metadata(context, request):
+    metadata_report = MetadataReport(request)
+    return metadata_report.generate()
+
+
+def metadata_report_factory(context, request):
+    qs = QueryString(request)
+    specified_type = qs.get_one_value(
+            params=qs.get_type_filters()
+    )
+    if specified_type == 'Annotation':
+        return _get_annotation_metadata(context, request)
+    elif specified_type == 'PublicationData':
+        return _get_publicationdata_metadata(context, request)
+    else:
+        return _get_metadata(context, request)
```

Comment:
Are there tests for these?

Reply:
Still working on tests. Just PRed to give you a quick look.